enclave/core/sgx/td.c

```diff
@@ -169,6 +169,28 @@ oe_sgx_td_t* oe_sgx_get_td()
     return td;
 }
 
+/*
+**==============================================================================
+**
+** oe_sgx_clear_td_states()
+**
+**     Internal API that allows an enclave to clear the td states.
+**
+**==============================================================================
+*/
+
+void oe_sgx_clear_td_states(oe_sgx_td_t* td)
```

Comment:
Should this be added to td_init instead? I'm undecided.

Can you also double-check if other fields can also be cleared.

Reply:
Based on current state machine implementation, we cannot clear the states in `td_init`. Here is the analysis
- We check the state before invoking `_handle_ecall`, which then invokes `td_init` when it's not initialized yet. As a result, it's too late to clear the states at `td_init`.
- However, it's okay to not clearing the states at `td_init` as we already guarantee the td page to be zeroed at enclave creation.


The reason for clearing the states is about a corner case (refer to the updated test case) where
1. The exception handler does not return but continuing the ecall context instead. 
2. The state will remaining `SECOND_LEVEL_EXCEPTION_HANDLING` after the ecall exit. Such logic was originally designed for the exception handler to make OCALLs. 
3. The host makes another ECALL that reuses the TCS, the new ECALL will have the state `SECOND_LEVEL_EXCEPTION_HANDLING`. 
4. Then the state check before  `_handle_ecall` will fail as it expects the state to be `ENTERED`.

Now the PR is updated to explicitly clearing the td states if the state is not `RUNNING`. I also added comments on why we cannot clear the td states on every ECALL returns (OE runtime does not know how the TCS is mapped to host threads). The upper-level runtime (e.g., mystikos) that implements thread support should be responsible for clearing the tds on thread exit.